README.md
# Grid Horizontal (grid-horizontal)
Horizontal layout library. Spreads elements with defined dimensions (i.e. images) in container with horizontal priority.
Requires jQuery.

## Screenshots
![Grid Horizontal Demo Image](http://igor.migasiewicz.pl/jpg/gh_git1.jpg)

## Notice
The script has been created during JavaScript and jQuery learning process. Both logic and syntax were not optimized enough yet, therefore consider it as beta version. Still lots to do..

## Current features
- setting maximum row height
- setting optional minimum container width for starting calculation (if smaller, elements are displayed vertically in single column)
- gutters
- hide last elements which cannot create proper row

## Installation
HTML (example):
```html
<div class="container" style="position: relative">
  <div class="item">
    <a href="#">
      <img class="cover" style="height: 100%" src="default.jpg" />
    </a>
  </div>
  <div class="item" style="width:200px; height: 100px;">
    Example content
    <img alt="Default UI image" src="defaultUIimg.jpg" />
  </div>
  <!-- more items -->
</div>
```
JS + jQuery:
```javascript
$('div.container').GridHorizontal({
	item: '.item',
	imageClass: '.cover',
	minWidth: 400,
	maxRowHeight: 350,
	gutter: 15,
	hideOverload: false,
})
```


CHANGELOG
CHANGELOG

-------------------------
v0.9.0
- creates horizontal layout based on container dynamic width and maximal possible row height
- creates gutters
- displays gallery vertically if container width is smaller than defined minimal value
-------------------------
v0.9.1
- bugfix: unexpexted massive bottom offset does not show anymore
- code optimization


archive/GridHorizontal_v0.9.1.js
/*
 * Grid-Horizontal v0.9.1
 */

//
// Extend jQuery
//
(function($) {
	$.fn.extend( {
	    GridHorizontal: function(object={}) {
	    	GridH(this, object);
	    	return this;
	    }
	});
})(jQuery);

//
// Grid-Horizontal
//

let GridH = function (container, settings = {}) {	

	// LAYOUT SETTINGS

	// Item selector
	let itemSelector = settings.item || '.item';

	// Gutter width
	let gutter = settings.gutter || 20;

	// Minimum container width to initiate calculation
	// If container is smaller than 'minRowWidth' value, then images spread vertically
	let minContainerWidth = settings.minWidth || 300;

	// Maximum row height
	// Script pulls next images as long as total row height gets smaller then 'maxRowHeight' value
	let maxRowHeight = settings.maxRowHeight || 320;

	//
	// SCRIPT
	//

	let containerWidth = container.width();
	let $items = container.children('.item');
	let $images = $items.find('img');
	
	let currentOffsetTop = 0;
	let currentOffsetLeft = 0;
	let finalContainerHeight = 0;

	// Start

	if(['absolute', 'fixed', 'relative'].includes(container.css('position')) === false ) {
		container.css('position', 'relative');
	}

	let imagesModified = minimizeImagesInfo($images);
	let rows = createRows(imagesModified);
	$(container).css('height', finalContainerHeight);
	
	pushToHTML($items, $images, imagesModified);

	// Destroy currents
	currentOffsetTop = 0;
	currentOffsetLeft = 0;

	// Stop

	//
	// FUNCTIONS
	//

	function minimizeImagesInfo(images)	{

		let array = [];
		for(let i=0;i<images.length;++i) {
			array[i] = {
				naturalWidth: images[i].naturalWidth,
				naturalHeight: images[i].naturalHeight,
				src: images[i].src,
				ratio: images[i].naturalWidth/images[i].naturalHeight,
			};
		}

		return array;
	}

	function createRows(array) {

		let rows = [];
		let currentRowHeight = 0;

		for(let i = 1, countItems = array.length; i <= countItems; i++) {			
			
			subArray = array.slice(0,i);
			
			if(containerWidth < minContainerWidth) {
				pushDimensionsByWidth(subArray, containerWidth, currentOffsetTop);
				currentRowHeight = getHeightByWidth(subArray[0], containerWidth);	
			} else if((currentRowHeight=sliceHeightCalc(subArray)) < maxRowHeight) {
				pushDimensionsByHeight(subArray, currentRowHeight, currentOffsetTop);				
			} else {
				continue;
			}

			increaseCurrentOffsetTop(currentRowHeight + gutter);

			array = array.slice(i);			
			countItems=array.length;
			i = 0;
		}

		if(array.length > 0) {
			pushDimensionsByHeight(array, maxRowHeight, currentOffsetTop);
			increaseCurrentOffsetTop(maxRowHeight);
		}

		finalContainerHeight = currentOffsetTop;

		return rows;
	}

	function pushToHTML(jqueryItems, jqueryImages, images) {

		for(let i=0;i<jqueryImages.length;++i) {
			if(jqueryImages[i].src != images[i].src) continue;

			$(jqueryItems[i]).css('position', 'absolute')
				.width(images[i].width)
				.height(images[i].height)
				//.offset({top: images[i].offsetTop, left: images[i].offsetLeft});
				.css('top', images[i].offsetTop)
				.css('left', images[i].offsetLeft)

			$(jqueryImages[i]).width(images[i].width)
				.height(images[i].height);
		} 
	}

	function sliceHeightCalc(array) {

		let calcHeight;
		let iGutter = array.length-1;
		let workWidth = containerWidth - iGutter*gutter;
		let sumRatio = 0;

		for (i=0,len=array.length; i<len;i++) {
			sumRatio += array[i].ratio;
		}
		calcHeight = (sumRatio > 0) ? workWidth/sumRatio : 0;

		return calcHeight;
	}

	function pushDimensionsByHeight(array, height, offset) {

		let currentOffsetLeft = 0;

		for(let i=0, len = array.length; i<len; i++) {
			item = array[i];

			item.width = height * item.ratio;
			item.height = height;
			item.offsetTop = offset;
			item.offsetLeft = currentOffsetLeft;

			currentOffsetLeft += item.width + gutter;
		}
	}

	function pushDimensionsByWidth(array, width, offset) {

		let item = array[0];

		item.width = width;
		item.height = getHeightByWidth(item, width);
		item.offsetTop = offset;
		item.offsetLeft = 0;
	}

	function increaseCurrentOffsetTop(value) {

		currentOffsetTop += value;
	}

	function getWidthByHeight(object, dimension) {
		return dimension * object.ratio;
	}

	function getHeightByWidth(object, dimension) {
		return dimension / object.ratio;
	}
}

GridHorizontal.js
/*
 * Grid-Horizontal v0.9.2
 * test
 */

//
// Extend jQuery
//

/*
CHANGELOG
- USE STRICT

*/
'use strict';

(function($) {
	$.fn.extend( {
	    GridHorizontal: function(object={}) {
	    	GridH(this, object);
	    	return this;
	    }
	});
})(jQuery);

//
// Grid-Horizontal
//

let GridH = function ($container, settings = {}) {	

	// LAYOUT SETTINGS

	// Item selector
	let itemSelector = settings.item || '.item';

	// Image selector
	let imageSelector = settings.imageClass || '.cover';

	// Gutter width
	let gutter = settings.gutter || 20;

	// Minimum container width to initiate calculation
	// If container is smaller than 'minRowWidth' value, then images spread vertically
	let minContainerWidth = settings.minWidth || 300;

	// Maximum row height
	// Script pulls next images as long as total row height gets smaller then 'maxRowHeight' value
	let maxRowHeight = settings.maxRowHeight || 320;

	// Hide overload
	// When true script won't show additional elements in the end of the list, those that are unable to form full row
	let hideOverload = settings.hideOverload || false;

	//
	// SCRIPT
	//
	
	let $items = $container.children('.item');
	let $images = $items.find('img');

	let containerWidth = $container.width();
	let currentOffsetTop = 0;
	let currentOffsetLeft = 0;
	let finalContainerHeight = 0;
	
	// Start

	// Define size and ratio of each element
	$items.each(function(i,elem) {
		let params = {};
		let $item = $(elem);
		let $currentImage = $item.find('img');

		if ($currentImage.length) {
			params.originalWidth = $currentImage[0].naturalWidth;
			params.originalHeight = $currentImage[0].naturalHeight;
		} else {
			params.originalWidth = parseInt($item.css('width'), 10);
			params.originalHeight = parseInt($item.css('height'), 10);
		}
		params.ratio = params.originalWidth/params.originalHeight;
		$item[0].GH = params;
	})

	// Prepare container for proper items display
	if(['absolute', 'fixed', 'relative'].includes($container.css('position')) === false ) {
		$container.css('position', 'relative');
	}
	
	calculateGrid($items);
	$container.css('height', finalContainerHeight);
	pushToHTML($items);

	// Destroy currents
	currentOffsetTop = 0;
	currentOffsetLeft = 0;

	// Stop

	//
	// FUNCTIONS
	//

	function calculateGrid(array) {

		let subArray;
		let currentRowHeight = 0;

		for(let i=1, countItems = array.length; i <= countItems; i++) {						
			subArray = array.slice(0,i);

			if(containerWidth < minContainerWidth) {
				let item = subArray[0];
				setGHParams(
					item, 
					containerWidth, 
					getHeightByWidth(containerWidth, item.GH.ratio), 
					currentOffsetTop, 
					0,
				);
				currentRowHeight = item.GH.height;	
			} else if((currentRowHeight=sliceHeightCalc(subArray)) < maxRowHeight) {
				let currentOffsetLeft = 0;

				for(let i=0, len = array.length; i<len; i++) {
					let item = array[i];
					setGHParams(
						item, 
						getWidthByHeight(currentRowHeight, item.GH.ratio), 
						currentRowHeight, 
						currentOffsetTop, 
						currentOffsetLeft,
					);
					currentOffsetLeft += item.GH.width + gutter;
				}
			} else {
				continue;
			}

			increaseCurrentOffsetTop(currentRowHeight + gutter);

			array = array.slice(i);			
			countItems=array.length;
			i = 0;
		}

		if(array.length > 0) {
			
			if(hideOverload == true) {
				array.each(function(i,elem) {
					$(elem).hide();
				});
			} else {
				for(let i=0, len = array.length; i<len; i++) {
					let item = array[i];
					setGHParams(
						item, 
						getWidthByHeight(maxRowHeight, item.GH.ratio), 
						maxRowHeight, 
						currentOffsetTop, 
						currentOffsetLeft,
					);
					currentOffsetLeft += item.width + gutter;
				}
				increaseCurrentOffsetTop(maxRowHeight);
			}
		}

		finalContainerHeight = currentOffsetTop;
	}

	function setGHParams(item, width, height, offTop, offLeft) {
		item.GH.width = width;
		item.GH.height = height;
		item.GH.offsetTop = offTop;
		item.GH.offsetLeft = offLeft;
	}

	function pushToHTML($items) {
		for(let i=0;i<$items.length;++i) {			
			let $itemLib = $items[i].GH;
			$($items[i]).css('position', 'absolute')
				.width($itemLib.width)
				.height($itemLib.height)
				.css('top', $itemLib.offsetTop)
				.css('left', $itemLib.offsetLeft)
		} 
	}

	function sliceHeightCalc(array) {
		let calcHeight;
		let workWidth = containerWidth - (array.length-1)*gutter;
		let sumRatio = 0;

		for (let i=0,len=array.length; i<len;i++) {
			sumRatio += array[i].GH.ratio;
		}
		calcHeight = (sumRatio > 0) ? workWidth/sumRatio : 0;

		return calcHeight;
	}

	function increaseCurrentOffsetTop(value) {
		currentOffsetTop += value;
	}

	function getWidthByHeight(dim, ratio) {
		return dim * ratio;
	}

	function getHeightByWidth(dim, ratio) {
		return dim / ratio;
	}
}
